lib/asciitable: implement `DiscardEmpty()`

jtagcat opened the pull request:
Before:
![image](https://user-images.githubusercontent.com/38327267/175877556-ab674491-764b-45cb-8e36-140929df2995.png)



After (`DiscardEmpty()`):
![image](https://user-images.githubusercontent.com/38327267/175877309-19a59119-38e1-4074-a116-c6026cef7d54.png)

Commit:
lib/asciitable: implement DiscardEmpty()

zmb3 commented on lib/asciitable/table.go:
```diff
@@ -41,6 +42,7 @@ type Table struct {
 	columns   []Column
 	rows      [][]string
 	footnotes map[string]string
+	mu        sync.Mutex
```
Why do we need a mutex here?

jtagcat replied:
The table is first searched, and then manipulated. Calling DiscardEmpty() and Add() or AsBuffer() at the same time from different goroutines panics, or eats data.

Though, it would probably be a bug by library user anyway.

zmb3 commented on lib/asciitable/table_test.go:
```diff
@@ -33,6 +33,13 @@ const headlessTable = `one  two
 
+const discardedTable = `one          two          three        
+---          ---          -----        
+%!v(MISSING) %!v(MISSING) %!v(MISSING) 
```
I'm not sure I understand this test. `%!v(MISSING)` is not something I would ever expect a user to see, why are we asserting that this is the output?

jtagcat replied:
I don't remember much, but below I had `testing for OOB panics` as a comment for this.

The code loops over the tables and manipulates them. It did panic at some point, and I'd say it's rather easy to make a mistake.

zmb3 replied:
Sure, it may be easy to make a mistake, but `%!v(MISSING)` is not a user friendly message. I would not want a user to ever see that. It looks like this code under some circumstances will present this to a user, so as-is I would not merge this.

jtagcat replied:
The PR does not affect `%!v(MISSING)` being printed.

![image](https://user-images.githubusercontent.com/38327267/192043108-bbc1ac82-6eb4-403f-b7d6-9bf363c13ce2.png)

It is caused by [`fmt.Fprint` in (*table).AsBuffer()](https://github.com/gravitational/teleport/blob/45c065aceed664f747532c7ca9463225b86079b9/lib/asciitable/table.go#L188).

Using the same test, template is `"%v\t%v\t%v\t%v\t%v\t%v\t"` and for the first line, `rowi` is nil (empty). The standard library prints `%!v(MISSING)` when there is no variable to populate the template from.

This may be addressed in a future change. Code for reproducing missing:
```go
func TestMissing(t *testing.T) {
	table := MakeTable([]string{"notmissing", "missing"})
	table.AddRow([]string{"ok"})
    result := table.AsBuffer().String() // row 1 will be "ok %!v(MISSING)"
}
```

zmb3 replied:
Right, I understand _why_ `%!v(MISSING)` could happen, but I don't think we need a test to verify that behavior, since it's more of a bug and not expected behavior, and Go could even decide to change the way missing format verbs are printed which would break the test.

jtagcat replied:
I added a commit. It should explain more. You are welcome to accept the PR with or without the last commit. I personally suggest without it.

Commit:
embed sync.Mutex instead of 'mu'

Commit:
lib/asciitable: remove test for panic in library

As requested by code review.
As a result, nil table rows are no longer guaranteed to not panic.

jtagcat commented:
@zmb3 any reasoning for closing?
